### PR TITLE
Add TeamsCallAdapter interface

### PR DESCRIPTION
**Pull request by PorterNan**
Add TeamsCallAdapter interface
Add createAzureCommunicationTeamsCallAdapter() function 
Extract common interface between teamsCallAdapter and CallAdapter

# What

# Why

# How Tested

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts [hunk 1]
```diff
@@ -369,12 +370,17 @@ export interface CallWithChatAdapterSubscriptions {
...
+export interface CallWithChatAdapterCommon
   extends CallWithChatAdapterManagement,
     AdapterState<CallWithChatAdapterState>,
     Disposable,
     CallWithChatAdapterSubscriptions {}
 
+/**
+ * @public
+ */
+export type CallWithChatAdapter = Omit<CallWithChatAdapterCommon, keyof ACSCallManagement> & ACSCallManagement;
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:382]: note: ARB is quite against the use of Omit and non-flat structures for public types, is there another way this could be defined?

Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:382]: (we might need a seperate work item to define a flat structure for the adapter interface as it already extends a few things)

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts [hunk 2]
```diff
@@ -65,7 +66,7 @@ export interface CallWithChatAdapterManagement {
    *
    * @public
    */
-  joinCall(microphoneOn?: boolean): Call | undefined;
+  joinCall(microphoneOn?: boolean): void;
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:69]: this looks like a breaking change

File: packages/communication-react/review/beta/communication-react.api.md [hunk 3]
```diff
@@ -77,6 +79,12 @@ import { VideoDeviceInfo } from '@azure/communication-calling';
 import { VideoStreamRenderer } from '@azure/communication-calling';
 import { VideoStreamRendererView } from '@azure/communication-calling';
 
+// @public (undocumented)
+export type ACSCallManagement = {
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:83]: Is there a reason to define seperate types for ACSCallManagement and TeamsCallManagement or can we use a generic?
```
interface CallManagement<CallType> {
    joinCall(microphoneOn?: boolean): <CallType> | undefined;
    startCall(participants: string[], options?: StartCallOptions): <CallType> | undefined;
};
```

File: packages/communication-react/review/beta/communication-react.api.md [hunk 4]
```diff
@@ -230,6 +237,10 @@ export type CallAdapterClientState = {
     alternateCallerId?: string;
 };
 
+// @public
+export interface CallAdapterCommon extends AdapterState<CallAdapterState>, Disposable, CallAdapterCallManagement, CallAdapterDeviceManagement, CallAdapterSubscribers {
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:241]: Can we avoid exporting the common interface and just publicly export `CallAdapter` and `TeamsCallAdapter` and use `CallAdapter | TeamsCallAdapter` where this is used? (in the public interface)

File: packages/communication-react/review/beta/communication-react.api.md [hunk 5]
```diff
@@ -527,7 +538,12 @@ export type CallErrors = {
...
+
+// Warning: (ae-internal-missing-underscore) The name "CallHandlersOf" should be prefixed with an underscore because the declaration is marked as @internal
+//
+// @internal (undocumented)
+export type CallHandlersOf<AgentType extends TeamsCallAgent | CallAgent> = AgentType extends CallAgent ? CallingHandlers : never | /* @conditional-compile-remove(teams-call) */ TeamsCallingHandlers;
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:546]: Public api file shouldn't have internals

File: packages/communication-react/review/beta/communication-react.api.md [hunk 6]
```diff
@@ -623,10 +644,19 @@ export interface CallState {
     startTime: Date;
     state: CallState_2;
     transcription: TranscriptionCallFeature;
+    type: 'Teams' | 'ACS';
 }
 
+// Warning: (ae-internal-missing-underscore) The name "CallTypeOf" should be prefixed with an underscore because the declaration is marked as @internal
+//
+// @internal (undocumented)
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:652]: Public api file shouldn't have internals

File: packages/communication-react/review/beta/communication-react.api.md [hunk 7]
```diff
@@ -239,9 +246,8 @@ export interface BrowserPermissionDeniedStyles extends BaseCustomStyles {
     troubleshootingLink?: ILinkStyles;
 }
 
-// @public
-export interface CallAdapter extends AdapterState<CallAdapterState>, Disposable, CallAdapterCallManagement, CallAdapterDeviceManagement, CallAdapterSubscribers {
-}
+// @public (undocumented)
+export type CallAdapter = Omit<CallAdapterCommon, keyof ACSCallManagement> & ACSCallManagement;
```
Comment (PorterNan) [inline packages/communication-react/review/beta/communication-react.api.md:250]: This Omit can be replaced by 'interface' method as we discussed @prprabhu-ms, will be working on it soon

File: packages/communication-react/review/beta/communication-react.api.md (hunk 4 above)
Comment (prprabhu-ms) [reply packages/communication-react/review/beta/communication-react.api.md:300]: Can't. `CallComposite` now only needs the `CallAdapterCommon` interface, not `CallAdapter`. This is because the composite only makes use of the common functionality.

File: packages/communication-react/review/stable/communication-react.api.md [hunk 8]
```diff
@@ -72,6 +72,12 @@ import { VideoDeviceInfo } from '@azure/communication-calling';
 import { VideoStreamRenderer } from '@azure/communication-calling';
 import { VideoStreamRendererView } from '@azure/communication-calling';
 
+// @public (undocumented)
+export type ACSCallManagement = {
```
Comment (prprabhu-ms) [inline packages/communication-react/review/stable/communication-react.api.md:76]: Don't use the acronym `ACS` in the API.

**Review by prprabhu-ms — COMMENTED** [review]
Delete packages/communication-react/temp copy/communication-react.api.json

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 9]
```diff
@@ -801,41 +881,40 @@ export const createAzureCommunicationCallAdapter = async ({
...
+    /* @conditional-compile-remove(PSTN-calls) */ alternateCallerId
+  });
+  const callAgent = await callClient.createTeamsCallAgent(credential, {
+    undefined
+  });
+  const adapter = createAzureCommunicationTeamsCallAdapterFromClient(callClient, callAgent, locator);
+  return adapter;
+};
+
+type AdapterType<CallType extends 'ACS' | 'Teams'> = CallType extends 'ACS'
```
[the remark this answers is not in the document]
Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:905]: Similar to above, can we define this type unconditionally to reduce the mental load of reading the code?

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 10]
```diff
@@ -925,6 +1028,93 @@ export const useAzureCommunicationCallAdapter = (
...
+  /**
+   * Optional callback called before the adapter is disposed.
+   *
+   * This is useful for clean up tasks, e.g., leaving any ongoing calls.
+   */
+  beforeDispose?: (adapter: CallAdapter) => Promise<void>
+): CallAdapter | undefined => {
+  return useAzureCommunicationCallAdapterGeneric(
+    args,
+    afterCreate as (adapter: CallAdapter) => Promise<CallAdapter>,
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1068]: This downcast is not safe. What if Contoso provides a callback that actually returns only a `Promise<CallAdapterCommon>`?

That said, I think we can keep the signature `afterCreate?: (adapter: CallAdapter) => Promise<CallAdapter>` because before this change, Contoso _must have_ used a callback with that signature (this PR only splits the definition of `CallAdapter` into `CallAdapterCommon` and `CallAdapter`) so that this downcast won't be necessary.

Same comment below.

**Review by prprabhu-ms — COMMENTED** [review]
Overall looks less controversial than the lower layers.

Thanks for working through this.

File: packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts [hunk 11]
```diff
@@ -60,7 +62,7 @@ export const END_CALL_PAGES: CallCompositePage[] = [
 ];
 
 /**
- * {@link CallAdapter} state for pure UI purposes.
+ * {@link CallAdapterCommon} state for pure UI purposes.
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts:65]: CallAdapter is used for ACS already, if we move it to AzureCommunicationCallAdapter, it would be a breaking type change :(
What do you think of CommonCallAdapter? Or AbstractCallAdapter?

File: packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts [hunk 12]
```diff
@@ -259,14 +261,16 @@ export interface CallAdapterCallManagement {
...
    */
-  startCall(participants: string[], options?: StartCallOptions): Call | undefined;
+  startCall(participants: string[], options?: StartCallOptions): void;
   /* @conditional-compile-remove(PSTN-calls) */
   /**
    * Start the call.
-   * @param participants - An array of {@link @azure/communication-common#CommunicationIdentifier} to be called
-   * @beta
+   *
+   * @param participants - An array of participant to join
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts:269]: Nice catch, it should be just a merge conflict

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 13]
```diff
@@ -238,13 +267,30 @@ export class AzureCommunicationCallAdapter implements CallAdapter {
...
-    this.handlers = createDefaultCallingHandlers(callClient, callAgent, deviceManager, undefined);
-
     this.onClientStateChange = onStateChange;
 
     this.subscribeDeviceManagerEvents();
 
     this.callClient.onStateChange(onStateChange);
+
+    /* @conditional-compile-remove(teams-identity-support) */
+    if (!_isACSCallAgent(callAgent)) {
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:277]: For 1&2, yes, I need to apply some patterns from last 2 PR reviews, just need some time after I finish interface change

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 14]
```diff
@@ -238,13 +267,30 @@ export class AzureCommunicationCallAdapter implements CallAdapter {
...
+      this.handlers = createDefaultTeamsCallingHandlers(
+        callClient,
+        callAgent,
+        deviceManager,
+        undefined
+      ) as CallHandlersOf<AgentType>;
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:283]: TypeScript is not smart enough to infer the type during the code closure :-(, the type would be unknown if I don't do the cast

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts (hunk 2 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:78]: This is for CommonCallAdapter - which is the interface(abstraction) needed by composite, it requires less so it is backward compatible 

For CallAdapter, it is remaining the same, joinCall still return Call | undefined there

**Comment by prprabhu-ms** [comment]
Please IM me when this is ready for another round of reviews. Would like to help get this over the finish line this week.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 10 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1068]: That's a point I wanted to discuss - lemme resolve all the other comments from Internal review and this PR, then we can chat about this a bit later
It is just because I found myself had a really hard time to implement a common function in contoso side

File: packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/CallAdapter.ts:65]: We chose CommonCallAdapter to align with handler layer in the end

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 15]
```diff
@@ -801,41 +881,40 @@ export const createAzureCommunicationCallAdapter = async ({
...
+};
+
+type AdapterType<CallType extends 'ACS' | 'Teams'> = CallType extends 'ACS'
+  ? CallAdapter
+  : /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAdapter;
+
+/**
+ * @private
+ */
+const useAzureCommunicationCallAdapterGeneric = <CallType extends 'ACS' | 'Teams' = 'ACS'>(
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:895]: This is a very smart one! I tried this and it works seamlessly. It also reduces overall code if-else branches, perfect!

**Comment by PorterNan** [comment]
Created a follow-up item to refactor implementation of adapter as we discussed @prprabhu-ms https://skype.visualstudio.com/SPOOL/_workitems/edit/2894791
Will pick it up a bit later once we finish main chunk of CTE

File: common/config/rush/pnpm-lock.yaml [hunk 16]
```diff
@@ -28,12 +28,19 @@ packages:
...
   /@azure/abort-controller/1.1.0:
     dependencies:
-      tslib: 2.4.1
+      tslib: 2.4.0
     dev: false
     engines:
       node: '>=12.0.0'
     resolution:
       integrity: sha512-TrRLIoSQVzfAJX9H1JeFjzAoDGcoK1IYX1UImfceTZpsyYfWr09Ss1aHW1y5TrrR3iq6RZLBwJ3E24uwPhwahw==
+  /@azure/communication-calling/1.8.0-beta.1:
```
Comment (JamesBurnside) [inline common/config/rush/pnpm-lock.yaml:37]: revert this (looks like the whole lockfile should be reverted)

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts (hunk 2 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:77]: Right, but technically anyone using the `CallWithChatAdapterManagement` interface directly will now hit a breaking change.
In actuality, ARB wants us to have flat interfaces anyway, perhaps (best in a separate PR if we do) we leave this interface as is, mark as deprecated and not used, then update the whole interface to be flat (internally we can still use common interfaces to ensure type correctness across different adapters)

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts [hunk 17]
```diff
@@ -386,12 +386,22 @@ export interface CallWithChatAdapterSubscriptions {
...
     CallWithChatAdapterSubscriptions {}
 
+/**
+ * @public
+ */
+export interface CallWithChatAdapter extends CallWithChatAdapterCommon {
+  joinCall(microphoneOn?: boolean): Call | undefined;
+  startCall(participants: string[], options?: StartCallOptions): Call | undefined;
+  /* @conditional-compile-remove(PSTN-calls) */
+  startCall(participants: CommunicationIdentifier[], options?: StartCallOptions): Call | undefined;
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:402]: just something I'm noticing here but not part of this PR, I didn't think you can have two overloaded functions with the same number of arguments. Should this not need to be:
```suggestion
  startCall(participants: CommunicationIdentifier[] | string[], options?: StartCallOptions): Call | undefined;
```

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 18]
```diff
@@ -954,6 +1055,24 @@ export const createAzureCommunicationCallAdapterFromClient: (
...
+ */
+export const createAzureCommunicationTeamsCallAdapterFromClient = async (
+  callClient: StatefulCallClient,
+  callAgent: TeamsCallAgent,
+  locator: CallAdapterLocator
+): Promise<TeamsCallAdapter> => {
+  const deviceManager = (await callClient.getDeviceManager()) as StatefulDeviceManager;
+  return new AzureCommunicationCallAdapter(callClient, locator, callAgent, deviceManager);
+};
+
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1075]: this file is now over 1000 lines, can we break up logic in this file into more isolated blocks in their own files

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 19]
```diff
@@ -350,28 +360,39 @@ export class AzureCommunicationCallAdapter implements CallAdapter {
...
-  private _joinCall(audioOptions: AudioOptions, videoOptions: VideoOptions): Call {
+  private _joinCall(audioOptions: AudioOptions, videoOptions: VideoOptions): CallTypeOf<AgentType> {
     const isTeamsMeeting = 'meetingLink' in this.locator;
     /* @conditional-compile-remove(rooms) */
     const isRoomsCall = 'roomId' in this.locator;
 
+    /* @conditional-compile-remove(teams-identity-support) */
+    if (_isTeamsCallAgent(this.callAgent)) {
+      if (isTeamsMeeting) {
+        return this.callAgent.join(this.locator as TeamsMeetingLinkLocator, {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:371]: Avoid this duplicate block which is the same as below. Instead invert the if:
```ts
if (_isTeamsCallAgent(this.callAgent) && !isTeamsMeeting) {
  throw new Error('Locator not supported by TeamsCallAgent');
}
...
if (isTeamsMeeting) {
  ...
```

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 20]
```diff
@@ -238,13 +248,13 @@ export class AzureCommunicationCallAdapter implements CallAdapter {
...
 
-    this.handlers = createDefaultCallingHandlers(callClient, callAgent, deviceManager, undefined);
-
     this.onClientStateChange = onStateChange;
 
     this.subscribeDeviceManagerEvents();
 
     this.callClient.onStateChange(onStateChange);
+
+    this.handlers = createHandlers(callClient, callAgent, deviceManager, undefined);
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:257]: did this move for a specific reason (i.e. this now _must_ be called after over functions)? If so add a comment that this must be called at the end

File: change-beta/@azure-communication-react-3ae22222-4a77-4173-80b3-ec236bef5e2f.json [hunk 21]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "Add CTE support to adapter layer",
```
Comment (prprabhu-ms) [inline change-beta/@azure-communication-react-3ae22222-4a77-4173-80b3-ec236bef5e2f.json:3]: Let's not use `CTE` in the changelog. Can this be more descriptive? `Add TeamsCallAdapter etc. to support composites powered by teams identity` or some such?

File: packages/communication-react/review/beta/communication-react.api.md [hunk 22]
```diff
@@ -242,7 +242,13 @@ export interface BrowserPermissionDeniedStyles extends BaseCustomStyles {
 }
 
 // @public
-export interface CallAdapter extends AdapterState<CallAdapterState>, Disposable, CallAdapterCallManagement, CallAdapterDeviceManagement, CallAdapterSubscribers {
+export interface CallAdapter extends CommonCallAdapter {
+    // (undocumented)
```
Comment (prprabhu-ms) [inline packages/communication-react/review/beta/communication-react.api.md:246]: All public API methods should be documented.

File: packages/communication-react/review/beta/communication-react.api.md [hunk 23]
```diff
@@ -689,8 +695,18 @@ export interface CallState {
     type: 'Teams' | 'ACS';
 }
 
+// @public (undocumented)
```
Comment (prprabhu-ms) [inline packages/communication-react/review/beta/communication-react.api.md:698]: Document this type and its methods.

File: packages/communication-react/review/beta/communication-react.api.md [hunk 24]
```diff
@@ -2956,6 +2982,16 @@ export interface SystemMessageCommon extends MessageCommon {
     messageType: 'system';
 }
 
+// @beta
+export interface TeamsCallAdapter extends CommonCallAdapter {
+    // (undocumented)
```
Comment (prprabhu-ms) [inline packages/communication-react/review/beta/communication-react.api.md:2987]: Document these methods.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 20 above)
Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:257]: I don't think this moved for any specific reason.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 25]
```diff
@@ -805,41 +826,38 @@ export const createAzureCommunicationCallAdapter = async ({
...
+  });
+  const adapter = createAzureCommunicationTeamsCallAdapterFromClient(callClient, callAgent, locator);
+  return adapter;
+};
+
+/**
+ * @private
+ */
+const useAzureCommunicationCallAdapterGeneric = <
+  AdapterType extends CallAdapter | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAdapter
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:854]: Implement this without conditional compilation. It is internal only, and the type `TeamsCallAdapter` is available here in both build flavors.

File: packages/react-composites/src/composites/CallComposite/adapter/createHandlers.ts [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+  _isACSCallAgent,
+  _isTeamsCall,
+  _isTeamsCallAgent
+} from '@internal/calling-stateful-client';
+
+/**
+ * @private
+ */
+export type CallHandlersOf<AgentType extends CallAgent | TeamsCallAgent> = AgentType extends CallAgent
+  ? CallingHandlers
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/createHandlers.ts:23]: Do you need this conditional compilation? Or can you set it up such that for stable build `TeamsCallingHandlers` is `never`?

**Review by prprabhu-ms — APPROVED** [review]
The elephant in the room is James' comment about a breaking API change. Otherwise LGTM.

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts (hunk 17 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:402]: I mentioned this in another PR for startCall as well, I don't remember the detail of this one, but this was what we end up with in the end, @anjulgarg please correct me if I am wrong

Comment (PorterNan) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts:402]: BTW, for "have two overloaded functions with the same number of arguments", we have some examples from "on" functions in calling sdk subscription. (But I still preferred "or" style than override if that is possible)

File: packages/react-composites/src/composites/CallComposite/adapter/createHandlers.ts (hunk 26 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/createHandlers.ts:23]: This is the only place to call TeamsCallingHandlers outside of their package, so I guess keeping it here would be nice to avoid adding another type into stable api as a never type (we will have to export TeamsCallingHandlers in communication-react)

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 20 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:271]: Previously it was because conditional compilation, but we extract the logic to createHandlers in the end, so I think it's good to move it back to the place

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 19 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:385]: Noticed there are 2 teams meeting condition here - one for CTE, one for regular ACS. So _isTeamsCallAgent(this.callAgent) still need to be there as a type guard, but I applied if(!isTeamsMeeting) {throw ""} to avoid an else closure :D

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 18 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1102]: Nice catch, @prprabhu-ms and me also discussed this yesterday and we decided to open another work item https://skype.visualstudio.com/SPOOL/_workitems/edit/3067003 for refactoring, because:
1. There is no impact on current api design
2. The refactor would be fairly large, which might involve more merge conflict for this PR (and this PR is already big enough)

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts (hunk 2 above)
Comment (PorterNan) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/CallWithChatAdapter.ts]: Although I thought there would be very few people directly use CallWithChatAdapterManagement, but I still need to say, you are right, let's not lie to semver number, this is a breaking change.

I tried flattening apis, but it turned out to be a big change for api, and flattening those apis is not the focus of this PR, so I prefer not to do it here in the PR.

The solution to make minimum change here is that I made some changes to current api - I had to add one more interface to isolate "joinCall" "startCall" functions, then once flattening happens, we just need to deprecate all of those together and we are good.

And good news here - we don't have to make api changes to CallWithChat anymore

File: packages/communication-react/review/beta/communication-react.api.md [hunk 27]
```diff
@@ -245,16 +245,28 @@ export interface BrowserPermissionDeniedStyles extends BaseCustomStyles {
...
+    startCall(participants: CommunicationIdentifier[], options?: StartCallOptions): Call | undefined;
 }
 
 // @public
 export type CallAdapterCallEndedEvent = {
     callId: string;
 };
 
+// @public @deprecated
+export interface CallAdapterCallManagement extends CallAdapterCallOperations {
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:261]: aha clever way to work around that! I like it!

File: packages/communication-react/review/beta/communication-react.api.md [hunk 28]
```diff
@@ -245,16 +245,28 @@ export interface BrowserPermissionDeniedStyles extends BaseCustomStyles {
...
+    joinCall(microphoneOn?: boolean): Call | undefined;
+    startCall(participants: string[], options?: StartCallOptions): Call | undefined;
+    // @beta
+    startCall(participants: CommunicationIdentifier[], options?: StartCallOptions): Call | undefined;
+}
+
 // @public
-export interface CallAdapterCallManagement {
+export interface CallAdapterCallOperations {
```
[the remark this answers is not in the document]
Comment (JamesBurnside) [reply packages/communication-react/review/beta/communication-react.api.md:269]: Note: we have a ton of these interfaces we currently export, would be great to go through mark them all as deprecated and just export the flat interfaces, but that's a big item of work but this use case highlights how it would reduce tech debt by allowing us to be flexible with interfaces that need only be internal-only

**Review by JamesBurnside — APPROVED** [review]
